### ISSUE TITLE
after upgrade to mobx4, @action.bound methods cannot be stub by Sinon

### ISSUE THREAD

**Issue by MastroLindus** [mobxjs/mobx#1398]
Hi,

I am in the process of updating our codebase from mobx3 to 4, including the tests that run on mocha+chai+sinon.

In some tests I noticed a change in behaviour of the @action.bound in combination with the Sinon stub method that prevents the method to be stub correctly:

```
import {stub} from "sinon";
import {action} from "mobx";
class Test {

constructor() {
this.normallyBoundMethod = this.normallyBoundMethod.bind(this);
}

@action.bound
public decoratedMethod(){
}


@action
public normallyBoundMethod(){
}

@action
public arrowMethod = () => {}
}

const instance = new Test();
stub(instance, "decoratedMethod");
stub(instance, "normallyBoundMethod");
stub(instance, "arrowMethod");
instance.decoratedMethod(); // doesn't get invoked in mobx3, does in mobx4
instance.normallyBoundMethod(); // doesn't get invoked in mobx3 or 4
instance.arrowMethod(); // doesn't get invoked in mobx3 or 4
```
(p.s.: we normally don't use the "arrowMethod form" with the @action decorator in our codebase, I am not even sure 100% if it works correctly as an action in that case, I just added it here as another example for binding).

In Mobx3 all the 3 forms of declaring a bound action can be stub using Sinon, therefore when they are invoked, the real function doesn't get invoked.
In Mobx4, normallyBoundMethod and arrowMethod can still be stub, but decoratedMethod that is decorated by @action.bound apparently gets invoked anyway (making our test fail).

If this reproduce on your side I am not really sure whether it's a bug,  or an undocumented breaking change.

**Comment by MastroLindus** [comment]
Looking more into this, I am noticed the following:
1)in some tests there are definitely some change in behaviour regarding this between mobx3 and mobx4, with no code changes, however:
2)the stub behaviour is also influenced if the stub property has been accessed before or after the actual stubbing:
```
console.log(instance.decoratedMethod);
stub(instance, "decoratedMethod");
```
works, while immediately stubbing doesn't.

**Comment by MastroLindus** [comment]
I appreciate the effort for the fix, however this is still broken in 4.1.0.

```
class MobxTest {

  @action.bound
  public decoratedMethod() {
    console.log("NOOOOOO");
  }

}

it.only("mobx stub test", () => {
    const instance = new MobxTest();
    stub(instance, "decoratedMethod");
    instance.decoratedMethod();
  });
```

this test will output the console.log, that is, the decoratedMethod gets executed instead of being stub.

However, referencing to the decoratedMethod before stubbing it will actually work:
```
it.only("mobx stub test", () => {
    const instance = new MobxTest();
    instance.decoratedMethod; // just a simple reference to the action method will make it work correctly
    stub(instance, "decoratedMethod");
    instance.decoratedMethod();
  });
```
This apparently fixes it. I can definitely use this workaround for my tests, however it's a behaviour is a bit deceiving and unpredictable

**Comment by mweststrate** [comment]
@MastroLindus yes, this is something we can't fix, (that is why it was already an issue in MobX 3). The problem is that `@action.bound` doesn't have access to the instance, it has access to the prototype. So how `action.bound` works is that upon the first actual _read_ (or write) it knows the `this`, and at that points it binds the function onto that `this`. However, as long you never read or write that property, the bound method will not be created and there is nothing being stubbed. That is why referencing the thing 'fixes' the issue. 

Note that for example the `@autobind` package has exactly the same issue: https://github.com/andreypopp/autobind-decorator/issues/47#issuecomment-366035041. 

Once the stage 3 proposals settles (that is, has a babel and TS implementation), we can fix this though as it will support this pattern :).

**Comment by ionutmiftode** [comment]
Hi @mweststrate ,
I have an issue related to this one. In mobx 3.x I was able to mock actions using jest something like 
`myStore.myAction = jest.fn()`. 
However after the update I receive the following error:
`TypeError: Cannot assign to read only property 'myAction' of object '#<MyStore>'`

it seems that in mobx 4.x actions are `writable: false` while in 3.x were `writable: false`.
I know that I can override them doing something like 
`Object.defineProperty(myStore, 'myAction', {value: jest.fn(), writable: true});`
but it will be a pain to change it in all the 100 failing tests :).

My question is if this issue will be also addressed or there is a configuration that I can make in order to make the actions writable?

**Comment by mweststrate** [comment]
A fix for that is already in master, will probably cut a release on monday

**Comment by mweststrate** [comment]
(and please, not that commenting on closed issues might not be noticed, I just accidentally ended up in this thread)

**Comment by mweststrate** [comment]
Should be fixed in 4.3.0
